src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,128 @@ pub struct ConfigSpace {
+    pub min_buffer_size: u32,
+    // An [`IoVecBufferMut`] covering all the memory we have available for receiving network
+    // frames.
+    pub iovec: IoVecBufferMut,
+    // A map of which part of the memory belongs to which `DescriptorChain` object
+    pub parsed_descriptors: VecDeque<ParsedDescriptorChain>,
+    // Buffers that we have used and they are ready to be given back to the guest.
+    pub deferred_descriptor: Option<ParsedDescriptorChain>,
+}
+
+impl RxBuffers {
+    /// Create a new [`RxBuffers`] object for storing guest memory for performing RX
+    fn new() -> Result<Self, IoVecError> {
+        Ok(Self {
+            min_buffer_size: 0,
+            iovec: IoVecBufferMut::new()?,
+            parsed_descriptors: VecDeque::with_capacity(FIRECRACKER_MAX_QUEUE_SIZE.into()),
+            deferred_descriptor: None,
+        })
+    }
+
+    /// Add a new `DescriptorChain` that we received from the RX queue in the buffer.
+    ///
+    /// SAFETY: The `DescriptorChain` cannot be referencing the same memory location as any other
+    /// `DescriptorChain`. (See also related comment in
+    /// [`IoVecBufferMut::append_descriptor_chain`]).
+    unsafe fn add_buffer(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        head: DescriptorChain,
+    ) -> Result<(), AddRxBufferError> {
+        let parsed_dc = self.iovec.append_descriptor_chain(mem, head)?;
+        if parsed_dc.length < self.min_buffer_size {
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+            return Err(AddRxBufferError::BufferTooSmall);
+        }
+        self.parsed_descriptors.push_back(parsed_dc);
+        Ok(())
+    }
+
+    /// Returns the number of available `iovec` objects.
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.iovec.len()
+    }
+
+    /// Mark the first `size` bytes of available memory as used.
+    ///
+    /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
+    /// means that `size` cannot be more than the currently available bytes in the underlying
+    /// `IoVecBufferMut`.
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+        // Since we were able to write a frame in guest memory, we should have at least one
+        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
+        // fundamentally broken.
+        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
+            "net: internal bug. Mismatch between written frame size and available descriptors",
+        );
+
+        // We are trying to mark as used more bytes than a single descriptor. However without
+        // VIRTIO_NET_F_MRG_RXBUF negotiated (we don't support it currently) the guest can't handle
+        // frames spanning multiple DescriptorChains. There is nothing we can do now, just pass the
+        // descriptor back to the guest without claiming any written bytes
+        if size > parsed_dc.length {
```

Comment:
Is there a situation when `size` is bigger than `parsed_dc.length`? When we write for mmds, the `write_all_..` will error out if we try to write more than the chain has and `readv` will always return value smaller equal to the chain capacity.

Reply:
Removing this `if` will also remove a need for `pop_front` inside the `drop_descriptor_chain` to count total capacity and it can be simplified to:
```Rust
    pub fn drop_descriptor_chain(&mut self, parse_descriptor: &ParsedDescriptorChain) {
        self.vecs.pop_front(parse_descriptor.nr_iovecs);
        self.len -= parse_descriptor.length as usize;
    }

```

Reply:
done